src/Enzyme.jl

```diff
@@ -1108,228 +1118,39 @@ Like [`gradient!`](@ref), except it using deferred mode.
     dx
 end
 
-"""
-    gradient(::ForwardMode, f, x; shadow=onehot(x))
-
-Compute the gradient of an array-input function `f` using forward mode. The
-optional keyword argument `shadow` is a vector of one-hot vectors of type `x`
-which are used to forward-propagate into the return. For performance reasons,
-this should be computed once, outside the call to `gradient`, rather than
-within this call.
-
-Example:
-
-```jldoctest
-f(x) = x[1]*x[2]
-
-grad = gradient(Forward, f, [2.0, 3.0])
-
-# output
-
-(3.0, 2.0)
-```
-"""
-@inline function gradient(::ForwardMode, f, x; shadow=onehot(x))
-    if length(shadow) == 0
-        return ()
-    end
-    res = values(only(autodiff(Forward, f, BatchDuplicatedNoNeed, BatchDuplicated(x, shadow))))
-    if x isa AbstractFloat
-        res[1]
-    else
-        res
-    end
+#NOTE: the following are to work around breaking changes but are inconsistent
+@inline jacobian_output_forward(df, df1, x) = gradient_output_forward(df, df1, x)
+@inline function jacobian_output_forward(df, df1::AbstractArray, x::AbstractArray)
+    reshape(reduce(hcat, df), gradient_output_size(df1, x))
 end
+@inline jacobian_output_forward(df, df1, x::AbstractArray) = copyto!(similar(x, eltype(df)), df)
 
-@inline function chunkedonehot(x, ::Val{chunk}) where chunk
-    sz = length(x)
-    num = ((sz + chunk - 1) ÷ chunk)
-    ntuple(Val(num)) do i
-        Base.@_inline_meta
-        onehot(x, (i-1)*chunk+1, i == num ? sz : (i*chunk) )
-    end
+@inline function jacobian(mode::ForwardMode, f, x; shadow=onehot(x))
+    #NOTE: these requires special handling to prevent breaking API
```

Comment:
this note doesn't apply here, no?

Reply:
It does, in the breaking version this is merely `gradient`, but in `main` currently there is a discrepancy between how `gradient` and `jacobian` handle outputs (owing to `gradient` sometimes returning tuples).  Therefore I had to make special accommodations to split the behavior here so it wouldn't break the API.

Reply:
Jacobian should also conditionally return tuples, for example, if inputs or outputs are not arrays/scalars

Reply:
This does that, as can be seen in the fallback where it returns the output of `derivative` `df` directly.  However I can only attest to the behavior being exactly the same in cases which are tested.  It was not entirely clear to me exactly what the semantics of the outputs are, one of the goals of the original PR was to help define them more clearly.

Reply:
Obviously likely undertested already, but since we're changing the implementation a bunch (and adding new fn's), would it be possible to add tests that confirm that all functions of the form input in (Float, Vector, Matrix, Struct) and output in (Float, Vector, Matrix, Struct)   work for as expected for (derivative, gradient, gradient!, jacobian) x ("", deferred)?

I realize this is kind of an obnoxious list, but also this is now the API surface (and presumably not sufficiently tested to start with).

If things don't break and work as intended, great!

If things break, we should fix it (or mark it as xfail, depending).

If things don't break, but are counter-intuitive, we should see if it would amount to a bugfix or breaking change for users, and fix it either immediately or follow up.

Nevertheless, for at least my sake reviewing it would be good to see this explicitly since the number of options is getting a bit unclear and large to think through all combinations (e.g. the old implementation explicitly had a switch for float vs vector vs matrix vs others that I'm now not as confident would be supported).

Reply:
I just added a bunch of tests from the other PR that cover part of this, but I still need to go through and comprehensively check for every method when I get a chance.  I already did some of this in testing on my end (not in the repo) so it shouldn't be too bad, though I'm probably missing some edge cases like different chunk sizes.